### DISCUSSION TITLE
Intellisense for (Powershell/Windows Terminal)

### DISCUSSION THREAD

**Discussion by osdenprog** [PowerShell/PowerShell#14986] [General]
I wonder why still till now we don’t have any nice **Intellisense** (Autocomplete) for terminals!? This idea is already done into previous version of **Poweshell_ISE** as per attached photo.

Why not to re-integrate this option again with this new (Windows Terminal) with the ability and option of extensibility to understand new _lexers_ for scripts and languages. It would be very helpful.
![PS-ise-0](https://user-images.githubusercontent.com/62112113/110654477-d612dc00-81c6-11eb-95f5-173906ced299.png)

**Comment by Ayanmullick** [comment]
@osdenprog , There is an enhancement issue for this. https://github.com/PowerShell/PSReadLine/issues/1362#issuecomment-601294373

**Reply by osdenprog** [reply]
_PSReadLine_ script was existing long time ago. I looked to the post suggested but still no graphical enhancement for a real **Intellisense** with a brief help as it was into **Powershell_ISE** 
Now with the **(Windows Terminal)** Built on **(electron)** technology, they can do it even much better, same like it is with autocomplete into **VSCode**

**Comment by vexx32** [comment]
The default PSRL key handler for menu-complete is Ctrl+Space, it gives pretty much this exact experience.

Personally I have it remapped to Tab, because I prefer it to normal tab complete.

**Comment by Amacc** [comment]
This might be helpful to you and might be in the ballpark of what you are looking for, the new psreadline beta has a prediction feature and tab autocomplete does pretty decent from cli experience.

Here is the prediction:
![image](https://user-images.githubusercontent.com/6657798/110733360-65ca8180-81eb-11eb-84a1-c5fd07d69775.png)

and with double tab for a list of all options (similar to a *nix system):
![image](https://user-images.githubusercontent.com/6657798/110733436-93172f80-81eb-11eb-9774-14deae83dfc2.png)

I set the following in my profile, if you use vs code its just `code $profile` and edit away.

```powershell
Set-PSReadlineKeyHandler -Key Tab -Function Complete
Set-PSReadlineOption -BellStyle None
Set-PSReadLineOption -PredictionSource History
```

And for the syntax yeah you have to perform a `get-help $command -syntax` to see what signatures it has.

**Reply by Ayanmullick** [reply]
@Amacc, One needs to run `Set-PSReadlineKeyHandler` every session, from one's profile. Is there any other way to persist `Set-PSReadlineKeyHandler` keys across sessions?

**Reply by osdenprog** [reply]
Thanks for all the replies, I understood all these details as I’m having long time experiences with such these things as a programmer. **_(PSReadLine)_** is introduced as a plug-ins just to beautify and colorize the textual Command line similarly to **(ZSH)** in linux. But still a textual interface. My post here is to suggest to powershell developers to integrate again the old **(Intellisense)** that was already done with **(Powershell_ISE)**. Similarly to what into the screenshot attached.

I wonder about Microsoft strategy sometime invent some nice and helpful then they abandon it without any reasons. As if someone is walking 1 step forward then 10 backward. Similarly is done with the development of new **(Powershell core)**. While there was so much more functionalities into old **(Powershell 5)** they abandon it in the new core version and some only some are moved to _(.net)_ with more complexity to use them.

With time they realize about some useful options and they re-integrate them like the option **(-ShowWindow)** that they return it back recently. Hopefully the same soon about the **(Intellisense)**

**Reply by vexx32** [reply]
What we had in the ISE isn't coming back. The ISE is a UI project, and PowerShell is fundamentally a shell application. The PowerShell team apparently never owned the ISE project itself, and they don't have the liberty to open source it from what I've heard.

They also don't appear to be wanting to integrate UI functionality into the shell itself; the majority of that effort has been put into PSReadLine and the VS Code extension. The PowerShell shell itself is not intended to ever be a full UI / IDE, so it's probably never going to have pretty UI menus like the ISE did.

**Reply by Amacc** [reply]
@Ayanmullick, So there is no way to integrate it directly into the powershell process itself, but you can easily save the powershell profile to perform these steps with no direct action on the users part. I created a repo with my profile and its contents and some of the folder around it [here](https://github.com/Amacc/PowershellProfile). Its been a minute since I have been in a remote shell but I would assume that it would have an environment initialization hook of some sort, so copying your remote host would do it, I can check into that use case later however.

@osdenprog Yeah the ISE isnt really a terminal, its a gui bolted onto a terminal, and IMO makes a terrible IDE. And I would agree its not a 1:1 replacement but a number of the use cases can be solved on the command line, intellisense is really a prediction use case, an autocomplete use case and a documentation/help man page use case rolled up into one. But if what you are looking for is the sanctioned supported platform its VS Code, it is the replacement to the ISE and allows the common conventions of a proper IDE. The powershell plugin is pretty good, not a lot of gripes with it, you get intellisense, debugging, remote execution, plugins and all the other niceities of a proper IDE.

**Reply by osdenprog** [reply]
I understand that (PowerShell) is a command line shell application. From other part **(Windows Terminal)** is a new terminal open source project from Microsoft relying on **_(Electron) UI_** & communicates under the hood with the (PowerShell) interface. 
It would be a nice suggestion if to integrate the **(Intellisese)** into the **(Windows Terminal)** but unfortunately WT teams, they don’t have any open forum discussion for suggesting new ideas at their page.
Moreover if they’ll integrate this option it’ll work also for the terminal of VS-CODE.

**Comment by gr4am** [comment]
{"app":"com.tencent.weather","desc":"","view":"RichInfoView","ver":"0.0.0.1","prompt":"","appID":"","sourceName":"","actionData":"","actionData_A":"","sourceUrl":"","meta":{"richinfo":{"adcode":"","city":"","air":"5","min":"-270","ts":"15158613","wind":"","date":"114514","max":"12138","type":"205"}},"text":"","sourceAd":"","extra":""}

**Comment by mardukbp** [comment]
You might want to check out [fig](https://fig.io/). Windows support is [under development](https://github.com/withfig/fig/issues/35). In the meantime you can use [PSfzf](https://github.com/kelleyma49/PSFzf), a PowerShell wrapper around the fuzzy finder fzf.

**Comment by osdenprog** [comment]
There is a very nice project called [CSharpREPL](https://github.com/waf/CSharpRepl) that integrated the same idea in Windows Terminal (WT), the same I was talking about years ago, the subject of this post page. But [CSharpREPL](https://github.com/waf/CSharpRepl) is only for C# and yet nothing the same done for PowerShell.

<img width="1350" height="581" alt="CSharpRepl" src="https://github.com/user-attachments/assets/6de630d4-5852-42e6-bfc5-febd29c20696" />


Yet, also [PSReadLine](https://github.com/PowerShell/psreadline) is still poor, not having the options that had it the legacy (**Powershell_ISE**) since the beginning.

There is also an interesting extension for (Visual Studio Code) Terminal, done for PowerShell that has somehow the advantages needed, but this run under the VS Code environment only. Why not do the same working for (WT)?
<img width="690" height="500" alt="VSCode Terminal" src="https://github.com/user-attachments/assets/88756d27-72b0-43f5-abbc-0346739e1f80" />

Also, a new similar project called [Inshellisense](https://github.com/microsoft/inshellisense) but still under development and not supporting PowerShell.

Here I’m showing the above examples to confirm that WT has all the abilities to extend it extensions wrappers to function similarly to (Powershell_ISE) with the IntelliSense ability and helpers for code autocompletion. I hope with the help of AI now, things are easier & coding is faster, so with whom are interested, we can collaborate to begin such project. Make a terminal rich & more intelligent. **Why the [WT team](https://github.com/microsoft/terminal) or [PowerShell Team](https://github.com/PowerShell/PowerShell) at @Microsoft not developed yet this Idea?**

**Reply by rhubarb-geek-nz** [reply]
> Why not do the same working for (WT)?

I would prefer to keep Windows Terminal to be a simple GUI console and do one thing well.

Not every program has to have Copilot integration, which is where it would end up. If VS code does what you want, use that. Folk talk about the demise of PowerShell ISE, why not treat VS Code as its replacement?

**Reply by mardukbp** [reply]
As CSharpREPL demonstrates, having auto-completion of PowerShell commands doesn't require any change to Windows Terminal. pwsh.exe could provide the same experience via a flag. I agree with you that Windows Terminal should not become a standalone version of the VS Code Terminal. But I think auto-completion is a feature that improves the UX a lot.

**Reply by osdenprog** [reply]
@rhubarb-geek-nz 
Each one has his preferences, for this the flexibility integrated in configuration of the apps. I’m talking about wrapper plugins or add-in, so whether you can add it or not. Similarly to PSReadLine. Even though, Now Microsoft ships PSReadLine with the PowerShell installation although still very primitive with options comparing to old legacy Pwershell_ISE

**Reply by rhubarb-geek-nz** [reply]
> Now Microsoft ships PSReadLine with the PowerShell installation although still very primitive with options comparing to old legacy Pwershell_ISE

Maybe, but compare like-with-like. PowerShell ISE was a GUI IDE, so a far better comparison with would be compare it to VS Code, which already has the plugins and integrations that you are asking for.

**Reply by osdenprog** [reply]
@rhubarb-geek-nz 
WT is also a GUI & has the extensibility option. To integrate on it the autocomplete with help will not transform it to IDE. At least to be more smart similarly to Linux terminals like Zsh or others.
I used Powershell_ISE as example only to explain better the similarity intent option needed to integrate it in WT PowerShell.

**Reply by rhubarb-geek-nz** [reply]
The boundary between PowerShell and its terminal is stdin, stdout and stderr using ANSI escape sequences. Are you saying that Windows Terminal should have it's own custom escape sequences to support this behaviour or is it something that the PowerShell PSReadLine module can do with any terminal with ANSI escape sequences and hence work on Linux, MacOS etc and over SSH?

**Comment by sharpchen** [comment]
Isn't this already possible by using things like https://github.com/PowerShell/CompletionPredictor with PSReadLine?
Somehow it doesn't complete commands but object members, might be a bug or non-goal(the readme is kind of confusing)? But it's possible to write your own predictor.

https://learn.microsoft.com/en-us/powershell/scripting/learn/shell/using-predictors?view=powershell-7.6
https://learn.microsoft.com/en-us/powershell/scripting/dev-cross-plat/create-cmdline-predictor?view=powershell-7.6

The custom predictor should be able to reproduce what vscode does for pwsh completion in terminal:
https://github.com/microsoft/vscode/blob/main/extensions/terminal-suggest/src/shell/pwsh.ts
